File: P2.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  4 11:36:06 2021

@author: jgdch
"""

# Time Complexity: O(M*N), M is the number of columns , N is the number of rows
#Space Complexity: O(1)
class Solution(object):
    def minFallingPathSum(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        for i in range(1,len(matrix)):
            for j in range(len(matrix[0])):
                if(j==0):
                    matrix[i][j]+=min(matrix[i-1][j],matrix[i-1][j+1])
                elif(j==(len(matrix[0])-1)):
                    matrix[i][j]+=min(matrix[i-1][j],matrix[i-1][j-1])
                else:
                    matrix[i][j]+=min(matrix[i-1][j-1],matrix[i-1][j],matrix[i-1][j+1])
        
        min_element=float("Inf")
        for i in range(len(matrix[0])):
            if(matrix[len(matrix)-1][i]<min_element):
                min_element=matrix[len(matrix)-1][i]
        return min_element
                
                    
        

File: p2.java
//time complesity : O(n*m)
//space complexity : O(n)

class Solution {
    public int minFallingPathSum(int[][] matrix) {

        if(nums.length == 0){
            return 0;
        }

        int n = matrix.length;
        int m = matrix.length[0];

        int[][] dp = new int[n][m];

        for(int i = 0; i < n; i++){
            dp[0][i] = matrix[0][i];
        }

        for( int i = 1; i < n ; i++){
            for( int j = 0; j < m; j++){
                if(j == 0){
                    dp[i][j] = matrix[i][j] + Math.min(dp[i - 1][j],dp[i - 1][j + 1]);

                }
                else if(j == n-1){
                    dp[i][j] = matrix [i][j] + Math.min(dp[i - 1][j],dp[i - 1][j - 1]);

                }
                else{
                    dp[i][j] = matrix [i][j] + Math.min(dp[i - 1][j - 1],Math.min(dp[i - 1][j],dp[i - 1][j + 1]));

                }
            }
        }
    }
}
